File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/ui/hero/adapter/HeroInfoAdapter.java
package com.akexorcist.repositoryarchcomponents.ui.hero.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.akexorcist.repositoryarchcomponents.R;
import com.akexorcist.repositoryarchcomponents.api.hero.response.Hero;

import java.util.List;

/**
 * Created by Akexorcist on 10/6/2017 AD.
 */

public class HeroInfoAdapter extends RecyclerView.Adapter<HeroInfoViewHolder> {
    private List<Hero> heroes;
    private HeroInfoListener heroInfoListener;

    public HeroInfoAdapter() {
    }

    public void setHeroes(List<Hero> heroes) {
        this.heroes = heroes;
    }

    @Override
    public HeroInfoViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.view_hero_info, parent, false);
        return new HeroInfoViewHolder(view);
    }

    @Override
    public void onBindViewHolder(HeroInfoViewHolder holder, int position) {
        Hero hero = heroes.get(position);
        holder.setName(hero.getName());
        holder.setAlias(hero.getAlias());
        holder.setType(hero.getType());
        holder.setImageUrl(hero.getUrl());
        holder.setCardClickListener(view -> onHeroSelected(hero));
    }

    private void onHeroSelected(Hero hero) {
        if (heroInfoListener != null) {
            heroInfoListener.onHeroSelected(hero);
        }
    }

    @Override
    public int getItemCount() {
        return heroes != null ? heroes.size() : 0;
    }

    public void setHeroInfoListener(HeroInfoListener heroInfoListener) {
        this.heroInfoListener = heroInfoListener;
    }

    public interface HeroInfoListener {
        void onHeroSelected(Hero hero);
    }
}

File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/ui/hero/adapter/HeroInfoAdapter.kt
package com.akexorcist.repositoryarchcomponents.ui.hero.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.akexorcist.repositoryarchcomponents.R
import com.akexorcist.repositoryarchcomponents.api.hero.respond.Hero

/**
 * Created by Akexorcist on 10/6/2017 AD.
 */

class HeroInfoAdapter : RecyclerView.Adapter<HeroInfoViewHolder>() {
    private var heroes: List<Hero>? = null
    private var heroInfoListener: ((hero: Hero) -> Unit)? = null

    fun setHeroes(heroes: List<Hero>) {
        this.heroes = heroes
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HeroInfoViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_hero_info, parent, false)
        return HeroInfoViewHolder(view)
    }

    override fun onBindViewHolder(holder: HeroInfoViewHolder, position: Int) {
        val hero = heroes!![position]
        holder.setName(hero.name)
        holder.setAlias(hero.alias)
        holder.setType(hero.type)
        holder.setImageUrl(hero.url)
        holder.setCardClickListener { onHeroSelected(hero) }
    }

    private fun onHeroSelected(hero: Hero) {
        heroInfoListener?.invoke(hero)
    }

    override fun getItemCount(): Int {
        return heroes?.size ?: 0
    }

    fun setHeroInfoListener(heroInfoListener: (hero: Hero) -> Unit) {
        this.heroInfoListener = heroInfoListener
    }

}
